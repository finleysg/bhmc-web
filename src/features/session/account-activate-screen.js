import "./session.scss"

import React from "react"

import { Button } from "components/buttons"
import { ErrorDisplay } from "components/errors"
import { Spinner } from "components/spinners"
import { useAuth } from "context/auth-context"
import { MdAccountCircle } from "react-icons/md"
import { useNavigate, useParams } from "react-router-dom"
import { useAsync } from "utils/use-async"

function AccountActivateScreen() {
  const { uid, token } = useParams()
  const { isError, isLoading, error, run } = useAsync()
  const { activate } = useAuth()
  const navigate = useNavigate()

  React.useEffect(() => {
    run(activate({ uid, token })).catch(() => {})
  }, [uid, token, run, activate])

  const handleLogin = () => {
    navigate("/session/login")
  }

  return (
    <div className="login">
      <div className="login__block active">
        <div className="login__header bg-green">
          <i>
            <MdAccountCircle />
          </i>
          Account Activation
        </div>
        {isLoading && (
          <div className="login__body">
            <h3 className="text-green">
              <i title="loading">
                <Spinner style={{ marginTop: ".6rem" }} />
              </i>
            </h3>
            <p>Activating your account...</p>
          </div>
        )}
        {!isLoading && isError && (
          <div className="login__body">
            <h3>Activation Failed</h3>
            <ErrorDisplay isError={isError} error={error} />
          </div>
        )}
        {!isLoading && !isError && (
          <div className="login__body">
            <h3>Your Account is Active</h3>
            <p>
              Thank you! Log in now to sign up for an event or update your profile. If you don't
              have a GHIN, you will receive one when you join us for the current golf season.
            </p>
            <Button onClick={handleLogin}>Login</Button>
          </div>
        )}
      </div>
    </div>
  )
}

export { AccountActivateScreen }
