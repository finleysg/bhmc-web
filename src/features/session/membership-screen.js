/** @jsxRuntime classic */
/** @jsx jsx */
import { jsx } from "@emotion/core"

import React from "react"

import { useClubEvent } from "hooks/event-hooks"
import ReactMarkdown from "react-markdown"
import { Link } from "react-router-dom"
import gfm from "remark-gfm"
import * as config from "utils/app-config"

function MembershipScreen() {
  const clubEvent = useClubEvent({ eventId: config.seasonEventId })

  return (
    <div className="content__inner">
      <header className="content__title">
        <h1>{config.currentSeason} Season Registration</h1>
      </header>
      <div className="row">
        <div className="col-lg-6">
          <div className="card">
            <div className="card-body">
              <h4 className="card-title">Membership Policies</h4>
              <React.Fragment>
                <h6 className="card-subtitle">Registration open: {clubEvent.signupWindow}</h6>
                <div className="card-text">
                  <ReactMarkdown source={clubEvent.notes} plugins={[gfm]} escapeHtml={true} />

                  <div className="col-12">
                    <p className="text-primary">
                      You need to have an account with us and be logged in to register for the{" "}
                      {config.currentSeason} season.
                    </p>
                  </div>
                </div>
                <div className="row">
                  <div className="col-12">
                    <Link
                      className="btn btn-primary"
                      style={{ marginRight: "1rem" }}
                      to="/session/login"
                    >
                      Login
                    </Link>
                    <Link className="btn btn-light" to="/session/account">
                      Sign Up For an Account
                    </Link>
                  </div>
                </div>
              </React.Fragment>
            </div>
          </div>
        </div>
        <div className="col-lg-6"></div>
      </div>
    </div>
  )
}

export default MembershipScreen
