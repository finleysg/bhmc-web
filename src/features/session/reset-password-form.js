import React from "react"

import { IconSubmitButton } from "components/buttons"
import { ErrorDisplay } from "components/errors"
import { FormGroup } from "components/forms"
import { Form, Formik } from "formik"
import { MdArrowForward } from "react-icons/md"
import { useAsync } from "utils/use-async"
import * as Yup from "yup"

function ResetPasswordForm({ onSubmit }) {
  const { isError, isLoading, error, run } = useAsync()

  function handleSubmit(values) {
    // any 400 or 500 is displayed to the user
    run(onSubmit(values)).catch(() => {})
  }

  return (
    <Formik
      initialValues={{ email: "" }}
      validationSchema={Yup.object({
        email: Yup.string().email("Invalid email address").required("A valid email is required"),
      })}
      onSubmit={(values) => handleSubmit(values)}
    >
      <Form>
        <FormGroup name="email" type="text" label="Email" />
        <IconSubmitButton loading={isLoading} color="blue">
          <MdArrowForward style={{ marginTop: ".6rem" }} />
        </IconSubmitButton>
        <ErrorDisplay isError={isError} error={error} />
      </Form>
    </Formik>
  )
}

export { ResetPasswordForm }
