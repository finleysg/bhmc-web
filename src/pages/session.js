import React from "react"

import { useLocation } from "react-router-dom"

function SessionScreen(props) {
  const location = useLocation()
  return (
    <div style={{ minHeight: "calc(100vh - 216px)" }}>
      <p>Session route {location.pathname}</p>
    </div>
  )
}

export { SessionScreen }
