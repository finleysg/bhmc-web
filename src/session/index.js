export { AccountActivateScreen } from "./account-activate-screen"
export { AccountConfirmScreen } from "./account-confirm-screen"
export { ResetPasswordScreen } from "./reset-password-screen"
export { LoginScreen } from "./login-screen"
export { RegisterScreen } from "./register-screen"
export { ResetPasswordSentScreen } from "./reset-password-sent-screen"
export { ResetPasswordConfirmScreen } from "./reset-password-confirm-screen"
export { ResetPasswordCompleteScreen } from "./reset-password-complete-screen"
