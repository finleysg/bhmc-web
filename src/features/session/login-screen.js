import "./session.scss"

import React from "react"

import { useAuth } from "context/auth-context"
import { LoginForm } from "features/session/login-form"
import { MdAccountCircle } from "react-icons/md"

function LoginScreen() {
  const { login } = useAuth()

  return (
    <div className="login">
      <div className="login__block active">
        <div className="login__header bg-green">
          <i>
            <MdAccountCircle />
          </i>
          Sign In to Your Account
        </div>

        <div className="login__body">
          <LoginForm onSubmit={login} />
        </div>
      </div>
    </div>
  )
}

export { LoginScreen }
