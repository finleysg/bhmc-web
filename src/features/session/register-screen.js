import "./session.scss"

import React from "react"

import { useAuth } from "context/auth-context"
import { RegisterForm } from "features/session/register-form"
import { MdAccountCircle } from "react-icons/md"

function RegisterScreen() {
  const { register } = useAuth()

  return (
    <div className="login">
      <div className="login__block active">
        <div className="login__header bg-blue">
          <i>
            <MdAccountCircle />
          </i>
          Create an Account
        </div>

        <div className="login__body">
          <RegisterForm onSubmit={register} />
        </div>
      </div>
    </div>
  )
}

export { RegisterScreen }
