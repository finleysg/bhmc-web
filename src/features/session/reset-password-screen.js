import "./session.scss"

import React from "react"

import { useAuth } from "context/auth-context"
import { MdAccountCircle } from "react-icons/md"

import { ResetPasswordForm } from "./reset-password-form"

function ResetPasswordScreen() {
  const { requestPasswordReset } = useAuth()

  return (
    <div className="login">
      <div className="login__block active">
        <div className="login__header bg-blue">
          <i>
            <MdAccountCircle />
          </i>
          Reset My Password
        </div>

        <div className="login__body">
          <p>
            Enter your email below. If you have an account with us with that email address, we will email you a link you
            can use to create a new password.
          </p>
          <ResetPasswordForm onSubmit={requestPasswordReset} />
        </div>
      </div>
    </div>
  )
}

export { ResetPasswordScreen }
