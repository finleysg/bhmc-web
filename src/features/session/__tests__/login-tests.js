import { act, render, screen, waitFor } from "@testing-library/react"
import userEvent from "@testing-library/user-event"

import React from "react"

import { LoginScreen } from "features/session/login-screen"
import { rest } from "msw"
import { setupServer } from "msw/node"
import { handlers } from "test/auth-handlers"
import { buildLoginForm } from "test/generate/auth"
import { AuthWrapper, deferred } from "test/test-utils"
import { authUrl } from "utils/client-utils"

const server = setupServer(...handlers)

const mockNav = jest.fn()
jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => mockNav,
}))

beforeAll(() => {
  server.listen()
})
afterAll(() => server.close())
afterEach(() => {
  server.resetHandlers()
  jest.clearAllMocks()
  // TODO: this is a useAuth implementation detail -- alternative?
  window.localStorage.removeItem("__bhmc_token__")
})

test("successful login", async () => {
  const { promise, resolve } = deferred()

  render(
    <AuthWrapper>
      <LoginScreen />
    </AuthWrapper>,
  )

  expect(screen.getByLabelText(/loading/i)).toBeInTheDocument()

  await act(() => {
    resolve()
    return promise
  })

  const { email, password } = buildLoginForm()

  userEvent.type(screen.getByRole("textbox", { name: /email/i }), email)
  userEvent.type(screen.getByLabelText(/password/i), password)
  userEvent.click(screen.getByRole("button"))

  await waitFor(() => expect(mockNav).toHaveBeenCalledWith("home"))
})

test("invalid credentials displays the error message", async () => {
  const testErrorMessage = ["Unable to log in with provided credentials."]
  const loginUrl = authUrl("token/login")
  server.use(
    rest.post(loginUrl, async (req, res, ctx) => {
      const err = ctx.json({ non_field_errors: testErrorMessage })
      return res(ctx.status(400), err)
    }),
  )

  const { promise, resolve } = deferred()

  render(
    <AuthWrapper>
      <LoginScreen />
    </AuthWrapper>,
  )

  expect(screen.getByLabelText(/loading/i)).toBeInTheDocument()

  await act(() => {
    resolve()
    return promise
  })

  const { email, password } = buildLoginForm()

  userEvent.type(screen.getByRole("textbox", { name: /email/i }), email)
  userEvent.type(screen.getByLabelText(/password/i), password)
  userEvent.click(screen.getByRole("button"))

  const alert = await screen.findByRole("alert")

  expect(alert.textContent).toMatchSnapshot()
})
