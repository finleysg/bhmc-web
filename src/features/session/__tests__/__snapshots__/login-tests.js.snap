// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`invalid credentials displays the error message 1`] = `"There was an error: Unable to log in with provided credentials."`;
